### PR TITLE
feat(RELEASE-987): add schema validator to pipelines via `check-data-keys task`

### PR DESCRIPTION
**Pull request by seanconroy2021**
This updates the `check-data-keys` task to now use a schema validator, ensuring data keys adhere to the specified schema.

Jira : [RELEASE-987](https://issues.redhat.com/browse/RELEASE-987)
Open PR for new dependency `check-jsonschema` in utils  [PR](https://github.com/konflux-ci/release-service-utils/pull/275)

### CODE REVIEW

File: tasks/check-data-keys/check-data-keys.yaml [hunk 1]
```diff
@@ -24,52 +29,52 @@ spec:
...
-            "cdn": [
-                "env"
-            ]
-        }'
+        if [ ! -f "/workspace/data/$(params.dataPath)" ]; then
+            echo "Data JSON file not found: /workspace/data/$(params.dataPath)"
+            exit 1
+        fi
+
+        if ! curl -s "$(params.schema)" -o /tmp/schema.json; then
```
Comment (mmalina) [inline tasks/check-data-keys/check-data-keys.yaml]: By default, curl will return 0 even if the response has code >400:
```
$curl -s https://raw.githubusercontent.com/seanconroy2021/release-service-catalog/refs/heads/RELEASE-987/bla
404: Not Found%
$ echo $?
0
```
So you'll need `--fail` or `--fail-with-body` - probably the latter, otherwise you will not know what failed.

So you'll have:
```suggestion
        if ! curl -s --fail-with-body "$(params.schema)" -o /tmp/output; then
          echo "Error: Failed to fetch schema from $(params.schema)"
          cat /tmp/output
          exit 1
        fi
```

File: tasks/check-data-keys/check-data-keys.yaml [hunk 2]
```diff
@@ -24,52 +29,52 @@ spec:
...
-                "topic",
-                "description",
-                "solution",
-                "references"
-            ],
-            "cdn": [
-                "env"
-            ]
-        }'
+        if [ ! -f "/workspace/data/$(params.dataPath)" ]; then
```
Comment (mmalina) [inline tasks/check-data-keys/check-data-keys.yaml]: Why was this changed from `$(workspaces.data.path)` to hardcoded `/workspace/data/`? That seems wrong.

File: tasks/check-data-keys/check-data-keys.yaml [hunk 3]
```diff
@@ -24,52 +29,52 @@ spec:
...
+
+        if ! curl -s "$(params.schema)" -o /tmp/schema.json; then
+            echo "Error: Failed to fetch schema from $(params.schema)"
+            exit 1
+        fi
 
-        if [ ! -f "$(workspaces.data.path)/$(params.dataPath)" ] ; then
-            echo "No data JSON was provided."
+        if ! jq empty /tmp/schema.json; then
+            echo "Error: Invalid JSON schema in /tmp/schema.json"
```
Comment (mmalina) [inline tasks/check-data-keys/check-data-keys.yaml]: I guess it's not really that important, since this script uses `set -x` so it will print out everything and you will see what `/tmp/schema.json`. But it seems wrong in principal to say that this hardcoded path has invalid json. Rather, it would be more appropriate to say that the json downloaded from $url is invalid.

File: tasks/check-data-keys/check-data-keys.yaml [hunk 4]
```diff
@@ -24,52 +29,52 @@ spec:
...
+        if [ $RC -ne 0 ]; then
+            exit 1
+        fi
 
-        for SYSTEM in "$@" ; do
-            if [[ $(jq -r ".$SYSTEM" <<< "$KEYS_JSON") == null ]] ; then
-                echo "Unsupported system value: $SYSTEM"
-                RC=1
-            fi
+        echo "Adding required fields to root for system: $*"
```
Comment (mmalina) [inline tasks/check-data-keys/check-data-keys.yaml]: I'm not sure about the direction this is going. In the Jira, I don't see anything about replacing the current behavior. From Jira it seems the goal was to just add a simple validation of the schema file (i.e. ensure that no unknown fields exist in the data json) and that's it. Has somebody told you otherwise? Or did you just come up with this?

If I'm reading this correctly, the suggested change would work as follows:

The schema will be augmented so that each _system_ is required and also all its fields are required. But that might not always be desirable. Even if you're required to defined releaseNotes, it doesn't automatically mean that you need to fill out all of the fields - that's why those required fields were listed in the `KEYS_JSON` variable.

In case of releaseNotes, I can see that `issues` was not in that list of required fields. But with the proposed change it would be. Also, `releaseNotes.content.images` was required to exist, but with this change we would no longer require `images` inside `content`.

So we should probably first discuss what we want out of this. But again, simply by looking at Jira, I would assume that the goal was just to add a simple schema validation on top of what we have here.

File: tasks/check-data-keys/check-data-keys.yaml (hunk 1 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I will add it in. Ty

File: tasks/.DS_Store
Comment (mmalina) [inline tasks/.DS_Store:1]: We definitely don't want this in the repo ;)

File: tasks/check-data-keys/check-data-keys.yaml [hunk 5]
```diff
@@ -55,21 +62,32 @@ spec:
...
 
-        exit $RC
+        if [ $RC -ne 0 ] ; then
+            exit 1
+        fi
+      
+        for system in "$@"; do
+          REQUIRED_KEYS=$(echo "$KEYS_JSON" | jq -c --arg system "$system" '.[$system]')
+
+          jq --arg system "$system" --argjson requiredKeys "$REQUIRED_KEYS" \
```
Comment (mmalina) [inline tasks/check-data-keys/check-data-keys.yaml]: But now you don't even check that `$system` is even in the data json, right? Like if it doesn't have `releaseNotes` at all - because that one is still optional.

File: schema/dataKeys.json [hunk 6]
```diff
@@ -319,7 +317,7 @@
           "type": "array",
           "items": {
             "type": "object",
-            "additionalProperties": false,
+            "additionalProperties": true,
```
Comment (mmalina) [inline schema/dataKeys.json]: This doesn't seem right. We don't want to allow any other fields other than the ones listed here.

File: tasks/check-data-keys/check-data-keys.yaml (hunk 1 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: With the new approach, we can now pass in the schema URL to the validator, which will handle the error if the JSON can't be found. 
` check-jsonschema --output-format=text --schemafile "$(params.schema)" "$(workspaces.data.path)/$(params.dataPath)"`

File: tasks/check-data-keys/check-data-keys.yaml (hunk 2 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Fixed

File: tasks/.DS_Store
Comment (seanconroy2021) [reply tasks/.DS_Store:1]: FIxed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 5 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Went back to the old way

File: schema/dataKeys.json (hunk 6 above)
Comment (seanconroy2021) [reply schema/dataKeys.json]: We have a problem with how one of the E2E  pipelines is defining source https://github.com/konflux-ci/e2e-tests/blob/b2ed0c117cb9ee726de051a3636f29e00eef6b7d/tests/release/pipelines/multiarch_advisories.go#L238C1-L242C8

So, it's just going to be a TMP change. Want see if we can pass e2e, and then we'll update depending on which one is correct.

File: tasks/check-data-keys/check-data-keys.yaml (hunk 4 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I have also kept the old branch so we can discuss it.

**Comment by johnbieren** [comment]
fyi @seanconroy2021 for github PRs, you don't have to change the title to `draft:` (I think gitlab might do this for you so maybe that is where you got the idea). At the top right where you click to add reviewers, you can just click convert to draft (or even better open the PR as a draft as then no reviewers get automatically added until you mark it as ready)

File: pipelines/push-disk-images-to-cdn/README.md [hunk 7]
```diff
@@ -20,6 +20,9 @@ Tekton Pipeline to push disk images to a cdn using pulp
 | taskGitUrl                      | The url to the git repo where the release-service-catalog tasks to be used are stored                  | Yes | https://github.com/konflux-ci/release-service-catalog.git |
 | taskGitRevision                 | The revision in the taskGitUrl repo to be used                                                         | No       | -               |
 
+## Changes in 1.2.0
+* Add new param `schema` to `check-dataz-keys` task to allow for a JSN schema to be passed in.
```
Comment (johnbieren) [inline pipelines/push-disk-images-to-cdn/README.md]: typo dataz

Comment (johnbieren) [reply pipelines/push-disk-images-to-cdn/README.md]: typo JSN

File: pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml [hunk 8]
```diff
@@ -191,6 +191,8 @@ spec:
       params:
         - name: dataPath
           value: "$(tasks.collect-data.results.data)"
+        - name: schema 
+          value: "https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/production/schema/dataKeys.json"
```
Comment (johnbieren) [inline pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml]: I think instead of hardcoding production, you want taskGitRevision

Comment (johnbieren) [reply pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml]: Although it may not make much difference. I don't see any easy way to use the taskGitUrl (since we want the raw version of the file) so e2e won't use the fork version anyways

File: pipelines/push-disk-images-to-cdn/README.md (hunk 7 above)
Comment (johnbieren) [reply pipelines/push-disk-images-to-cdn/README.md]: (These are in the other READMEs too)

File: tasks/check-data-keys/README.md [hunk 9]
```diff
@@ -14,6 +13,10 @@ Currently, `releaseNotes`, and `cdn` are the only supported systems.
 |----------|---------------------------------------------------------|----------|---------------|
 | dataPath | Path to the JSON string of the merged data to use       | No       |               |
 | systems  | The systems to check that all data keys are present for | Yes      | []            |
+| schema   | The URl to the schema                                   | Yes      | https://github.com/konflux-ci/release-service-catalog/blob/production/schema/dataKeys.json |
+
+## Changes in 10.0.0
```
Comment (johnbieren) [inline tasks/check-data-keys/README.md]: This should be 1.0.0 not 10.0.0

File: tasks/check-data-keys/check-data-keys.yaml [hunk 10]
```diff
@@ -4,17 +4,22 @@ kind: Task
 metadata:
   name: check-data-keys
   labels:
-    app.kubernetes.io/version: "0.9.2"
+    app.kubernetes.io/version: "0.10.0"
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: 1.0.0 please

File: tasks/check-data-keys/check-data-keys.yaml [hunk 11]
```diff
@@ -4,17 +4,22 @@ kind: Task
...
+    Tekton task to check that all required information is present to use the specified system(s) 
+    and the data is valid against the schema.
   params:
     - name: dataPath
       description: Path to the JSON string of the merged data to use
       type: string
+    - name: schema 
+      description: URL to the JSON schema file to validate the data against
+      type: string
+      default: https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/production/schema/dataKeys.json
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: I think I am okay with just having the default here like you have it and not passing it in the pipeline definitions. I don't think we can easily make it work for ci anyways

File: tasks/check-data-keys/check-data-keys.yaml [hunk 12]
```diff
@@ -24,7 +29,8 @@ spec:
       description: The workspace where the data JSON file resides
   steps:
     - name: check-data-keys
-      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
+    # TODO: Update the image to the latest release-service-utils image
+      image: quay.io/redhat-user-workloads/rhtap-release-2-tenant/release-service/release-service-utils:on-pr-403edb0827b86a46933608edc09a8b8598373f0a
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: You have a todo, so hopefully it lasts long enough for you, but fyi the redhat-user-workloads images can vanish at any time without warning, so it is probably better to build a copy of the utils image in your quay org and use that for testing

File: tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml [hunk 13]
```diff
@@ -0,0 +1,85 @@
+---
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: test-check-data-keys-fail-missing-schema-key
+  annotations:
+    test/assert-task-failure: "run-task"
+spec:
+  description: |
+
```
Comment (johnbieren) [inline tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml]: TODO

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Ok, so I have removed the schema parm in the pipelines

File: tasks/check-data-keys/check-data-keys.yaml [hunk 14]
```diff
@@ -24,52 +29,46 @@ spec:
...
+              ;;
+            cdn)
+              jq '.required += ["cdn"] | .properties.cdn.required += ["env"]' /tmp/output > /tmp/change
+              mv /tmp/change /tmp/output
+              ;;
+            *)
+              echo "Unsupported system value: $SYSTEM"
+              rc=1
+              ;;
+          esac
```
Comment (seanconroy2021) [inline tasks/check-data-keys/check-data-keys.yaml]: @johnbieren  This was my idea of how we make the data keys required  and then  allow the schema validator to catch the missing keys instead, e.g. 

```shell
[run-task : check-data-keys] Schema validation errors were encountered.
[run-task : check-data-keys]   /workspace/data/data.json::$.cdn: Additional properties are not allowed ('somekey' was unexpected)
[run-task : check-data-keys]   /workspace/data/data.json::$.cdn: 'env' is a required property
```

File: pipelines/push-disk-images-to-cdn/README.md (hunk 7 above)
Comment (seanconroy2021) [reply pipelines/push-disk-images-to-cdn/README.md]: Removed the pipeline see: https://github.com/konflux-ci/release-service-catalog/pull/603#discussion_r1801148486

File: tasks/check-data-keys/README.md (hunk 9 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/README.md]: Fixed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 10 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Fixed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 12 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I MR the Utils image today so it's updated now ...  I need to bump the images elsewhere then.

File: tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml (hunk 13 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml]: Added that in

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: That's fair based on what my comments were, but didn't we talk about in the call to have the value use taskGitRevision value? So then it would have to be passed in the pipelines

File: tasks/check-data-keys/README.md [hunk 15]
```diff
@@ -1,7 +1,6 @@
 # check-data-keys
 
-Tekton task to check that all required information is present in the data file in order to use the
-specified system(s). If any required keys are missing, the task will fail.
+Tekton task to check that all required information is present in the data file in order to use the specified system(s). If any required keys are missing, the task will fail. It will also validate against a JSON schema, ensuring all keys follow the correct format and are allowed.
```
Comment (johnbieren) [inline tasks/check-data-keys/README.md]: I think this can be reworded now as it isn't the existing check + the schema check now; it is just the schema check

File: tasks/check-data-keys/README.md [hunk 16]
```diff
@@ -14,6 +13,10 @@ Currently, `releaseNotes`, and `cdn` are the only supported systems.
 |----------|---------------------------------------------------------|----------|---------------|
 | dataPath | Path to the JSON string of the merged data to use       | No       |               |
 | systems  | The systems to check that all data keys are present for | Yes      | []            |
+| schema   | The URl to the schema                                   | Yes      | https://github.com/konflux-ci/release-service-catalog/blob/production/schema/dataKeys.json |
+
+## Changes in 1.0.0
+* Adding a schema validater to the end of the task
```
Comment (johnbieren) [inline tasks/check-data-keys/README.md]: More like replace the check with a schema validator

File: tasks/check-data-keys/tests/test-check-data-keys-fail-unsupported-system.yaml [hunk 17]
```diff
@@ -87,6 +88,7 @@ spec:
           value:
             - releaseNotes
             - unsupported
+            - random
```
Comment (johnbieren) [inline tasks/check-data-keys/tests/test-check-data-keys-fail-unsupported-system.yaml:93]: why? Isn't unsupported sufficient?

File: tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+              }
+              EOF
+    - name: run-task
+      taskRef:
+        name: check-data-keys
+      params:
+        - name: dataPath
+          value: "data.json"
+        - name: schema
+          value: "https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/production/schema/random-schema.json"
```
Comment (johnbieren) [inline tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml]: Can you make this a little more obvious and set it to like "non-existent-schema.json" instead of a url that looks like the real one :slightly_smiling_face:

File: tasks/check-data-keys/check-data-keys.yaml [hunk 19]
```diff
@@ -24,52 +29,46 @@ spec:
       description: The workspace where the data JSON file resides
   steps:
     - name: check-data-keys
-      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
+      image: quay.io/konflux-ci/release-service-utils@sha256:b8a4a4ff239d7789029a7bbd8c89d41453107d4391c1aaf7ea158f343d5624f6
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: We actually should use sha tags everywhere, but we don't we use the commit tag so can you keep it consistent?

File: tasks/check-data-keys/check-data-keys.yaml [hunk 20]
```diff
@@ -24,52 +29,46 @@ spec:
...
-            ]
-        }'
-
         if [ ! -f "$(workspaces.data.path)/$(params.dataPath)" ] ; then
             echo "No data JSON was provided."
             exit 1
         fi
 
-        RC=0
+        if ! curl -s --fail-with-body "$(params.schema)" -o /tmp/output ; then
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: Maybe put it in /tmp/schema or something more intuitive?

File: tasks/check-data-keys/check-data-keys.yaml [hunk 21]
```diff
@@ -24,52 +29,46 @@ spec:
...
-            for KEY in $(jq -r ".${SYSTEM}[]" <<< "$KEYS_JSON") ; do
-                if [[ $(jq ".$SYSTEM.$KEY" "$(workspaces.data.path)/$(params.dataPath)") == "null" ]] ; then
-                    echo "Missing required $SYSTEM key: $KEY"
-                    RC=1
-                fi
-            done
+        for SYSTEM in "$@" ; do
+          case "$SYSTEM" in
+            releaseNotes)
+              jq '.required += ["releaseNotes"] | .properties.releaseNotes.required += [
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: Is this necessary? It would be really nice to have the fields set as required or not in the schema instead of here and not have to maintain the required fields in two locations (this task + the schema)

File: tasks/check-data-keys/check-data-keys.yaml (hunk 20 above)
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: It'd be really nice in the mocks.sh of this task to mock curl with like `cat ../../../../schema.json` (not the real path) so that changes to the schema that are part of this PR are included in the pr unit tests

File: tasks/check-data-keys/README.md (hunk 15 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/README.md]: Reworded it.

File: tasks/check-data-keys/README.md (hunk 16 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/README.md]: Fixed

File: tasks/check-data-keys/tests/test-check-data-keys-fail-unsupported-system.yaml (hunk 17 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/tests/test-check-data-keys-fail-unsupported-system.yaml:93]: I wanted to check if a user had multiple unsupported systems that would output all of them, not just one.

File: tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml (hunk 18 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml]: fixed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 19 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I will create a ticket for it .... fixed it for time being

File: tasks/check-data-keys/check-data-keys.yaml (hunk 20 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: fixed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 21 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Ok, so now the systems passed in will be added to the data.json, and the schema will handle the logic.

File: tasks/check-data-keys/check-data-keys.yaml (hunk 20 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Will create it :)

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Yes, I can change it.  I am just scared it will be harder to pass e2e as the schema has been updated will be in the development branch.

File: tasks/check-data-keys/check-data-keys.yaml (hunk 4 above)
Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: Closing the loop here, Sean and I met about it and as long as his json schema can do the same things as the current code, the current code can be replaced. So, the new solution has to check the systems passed, type check the values in the system, and ensure that the required fields are present in the system. It seems Sean got this working with some if when stuff in the schema

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: But isn't the update to the schema just adding the required keys? So the merged one (that e2e will run with) has nothing required, so it will pass?

File: tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml (hunk 18 above)
Comment (johnbieren) [reply tasks/check-data-keys/tests/test-check-data-keys-fail-missing-schema.yaml]: I meant replace the entire value `https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/production/schema/non-existent-schema.json` with `non-existent-schema.json`, but this is fine too

File: tasks/check-data-keys/check-data-keys.yaml [hunk 22]
```diff
@@ -4,17 +4,23 @@ kind: Task
...
+    Tekton task to check that all required information is present to use the specified system(s) 
+    and the data is valid against the schema.
   params:
     - name: dataPath
       description: Path to the JSON string of the merged data to use
       type: string
+    - name: schema 
+      description: URL to the JSON schema file to validate the data against
+      type: string
+      # TODO: Update the default URL to the correct one
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: You could also update the schema in a separate PR that we merge first

File: tasks/check-data-keys/check-data-keys.yaml [hunk 23]
```diff
@@ -24,52 +30,26 @@ spec:
...
-                    echo "Missing required $SYSTEM key: $KEY"
-                    RC=1
-                fi
-            done
-        done
+        systemsJSON=$(echo "$@" | jq -R 'split(" ")')
 
-        exit $RC
+        jq --argjson systems "$systemsJSON" '.systems += $systems' \
+            "$(workspaces.data.path)/$(params.dataPath)" > "tmp/systems"
```
Comment (johnbieren) [inline tasks/check-data-keys/check-data-keys.yaml]: I think you should use `/tmp/systems` (missing leading /). I am not sure a tmp dir will exist in the homedir of the container

**Review by johnbieren — COMMENTED** [review]
```
[run-task : check-data-keys] Schema validation errors were encountered.
[run-task : check-data-keys]   /workspace/data/data.json::$.releaseNotes: 'product_id' is a required property
```
awesome!

Can you add a test where a value is not of the proper type? We had no type checking in this task before this PR, but now we do, so I'd like a test for it

File: tasks/check-data-keys/check-data-keys.yaml (hunk 4 above)
Comment (mmalina) [reply tasks/check-data-keys/check-data-keys.yaml]: OK, thanks for the update. I just looked briefly now, so let me see if my understanding is correct: The schema now has a new field `systems` which is not really part of the data file. But when the check-data-keys task is run, it will inject it there and that will make some of the field required and then the schema is validated. Is that how it works?

It seems a bit hacky, but fine I guess.

Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: That's how I understand it, yeah. It would be nice to not have the systems key in the schema at all, but it gets us everything else we want AFAICT. If we discover an improvement later on, great, but for now I think it is plenty good enough

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: So e2e has an extra key [here](https://github.com/konflux-ci/e2e-tests/blob/c69d596d294d58aee5aa270d38783cc7d613b5f7/tests/release/pipelines/multiarch_advisories.go#L238) then the product_id is defined as int in the schema and in the [docs](https://konflux-ci.dev/release-service-docs/reference/catalog/data-key-options/#:~:text=source%3A%20%3Cstring%3E-,product_id,-%3A%20%3Cinteger%3E) but in e2e its a [string](https://github.com/konflux-ci/e2e-tests/blob/c69d596d294d58aee5aa270d38783cc7d613b5f7/tests/release/pipelines/multiarch_advisories.go#L252) so it will fail for the time being.

So we  need to update the schema to include the source, git & URL
then we need to decide if we want to use string or int for product_id in ReleaseNotes

**Comment by seanconroy2021** [comment]
> ```
> [run-task : check-data-keys] Schema validation errors were encountered.
> [run-task : check-data-keys]   /workspace/data/data.json::$.releaseNotes: 'product_id' is a required property
> ```
> 
> awesome!
> 
> Can you add a test where a value is not of the proper type? We had no type checking in this task before this PR, but now we do, so I'd like a test for it

added one in

```shell
[run-task : check-data-keys] Schema validation errors were encountered.
[run-task : check-data-keys]   /workspace/data/data.json::$.releaseNotes.product_id: 'rh-123' is not of type 'integer'
```

File: tasks/check-data-keys/check-data-keys.yaml (hunk 23 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Fixed

File: tasks/check-data-keys/check-data-keys.yaml (hunk 22 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I think that would make more sense. WIll do it Monday

Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: https://github.com/konflux-ci/release-service-catalog/pull/630

File: tasks/check-data-keys/check-data-keys.yaml (hunk 11 above)
Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: https://github.com/konflux-ci/e2e-tests/blob/c69d596d294d58aee5aa270d38783cc7d613b5f7/tests/release/pipelines/multiarch_advisories.go#L238 is wrong and should be fixed
product_id I'd have to check some users to see what the real thing is. e2e could be wrong for that too

Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Ok, perfect then it should pass e2e, then 🙏 when the above is resolved

**Review by johnbieren — COMMENTED** [review]
All looks great to me, just need to get the other PR merged. We should pass the schema to the task in the pipelines (so we can use taskGitRevision). That should probably be part of this PR

**Comment by seanconroy2021** [comment]
ç

File: tasks/check-data-keys/check-data-keys.yaml (hunk 20 above)
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: I was thinking about this: Would it make more sense for the schema to have its own tests? As the schema can be updated, but the check-data-leys will not be changed, meaning the tests will not run. Also, since the schema is now in another PR, it won't pull the correct schema from local.

Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: I don't think it makes sense for a schema to have its own tests. https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-advisory/tests/pre-apply-task-hook.sh#L13 tells me that the mocks execute in a directory with the task. So, if you do some `ls` calls and `pwd`, I would bet you can find the path to the schema and then mock the curl to return it

Comment (mmalina) [reply tasks/check-data-keys/check-data-keys.yaml]: > tells me that the mocks execute in a directory with the task

Depends what you mean by mock execution. When `pre-apply-task-hook.sh` is run, you have access to files. So you inject a mock to the task script. But when the task runs (and a mock is hit), you don't have access to any files in the repo - you're in a container.

But I'm not really sure why a task test would need the real schema. A task check should verify that the task does what it should do - in this case it runs schema validation using the schema you supply. So to verify that, using some simple schema should be enough.

Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: It is nice to use the real schema because if a PR updates the schema in a way that breaks the test, we would catch it (assuming this task was also updated in the same PR). 
> So you inject a mock to the task script. But when the task runs (and a mock is hit), you don't have access to any files in the repo - you're in a container.

Yeah, so the mock can be like
```
curl() {
    cat ../../../../schema.json
}

Comment (mmalina) [reply tasks/check-data-keys/check-data-keys.yaml]: > Yeah, so the mock can be like

How will the file get to that location in that container that executes the task script in the test?

Comment (johnbieren) [reply tasks/check-data-keys/check-data-keys.yaml]: Yeah I talked about this in dms with Sean. If the loadstr thing doesn't end up working, I am fine with him scrapping my idea

Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Ok, I have added one that curls the development. Let me know if we are happy with or we can just remove it.

**Review by johnbieren — APPROVED** [review]
